src/s3.py

```diff
@@ -119,13 +121,13 @@ def save_ndarray_to_s3_as_npy(array: Any, s3_path: str) -> None:
     """Saves a NumPy ndarray to an S3 bucket as a .npy file."""
     bucket, key, s3client = validate_s3_pattern(s3_path)
 
-    # Convert the NumPy array to bytes
-    array_bytes = array.tobytes()
+    with tempfile.TemporaryFile() as f:
+        np.save(f, array)
+        f.seek(0)
 
-    # Upload the array bytes to S3
-    try:
-        s3client.put_object(Body=array_bytes, Bucket=bucket, Key=key)
-    except errors.NoSuchBucket:
-        raise ValueError(f"Bucket {bucket} does not exist")
-    except Exception as e:
-        raise e
+        try:
+            s3client.upload_fileobj(f, bucket, key)
```

Comment:
I've always thought that a `with` statement always created the scope of the variable- reading up this is not true.... you learn something new every day :) thanks @THOR300